### ISSUE TITLE
h264: unwrapping of None value

### ISSUE THREAD

**Issue by Gnurou** [chromeos/cros-codecs#31]
This unwrap in `src/decoder/stateless/h264.rs` seems to fail once in a while when playing around with ARCVM:

```
    /// Submits the picture to the accelerator.
    fn submit_picture(&mut self) -> Result<(PictureData, T), DecodeError> {
        let picture = self.cur_pic.take().unwrap();
```

**Comment by Gnurou** [comment]
This may be happening because the logic in `decode_access_unit` always expects to have a full picture in the input bitstreasm, which is not necessarily the case. We should probably change the logic to only process one NAL unit per call to `decode`, or at least to only call `submit_picture` if we indeed had a picture.

The other place where `submit_picture` is called is related to multi-field frames, and we should probably add some sanity checks there as well to not crash on invalid streams. Maybe we can pass the picture as an argument to `submit_picture`, which would force the caller to `take` it itself and to check that it actually exists.

**Comment by dwlsalmeida** [comment]
@Gnurou Hi Alex, perhaps it's time for me to test this on ARCVM as well, that will help with this and #30

Yes the code assumes an access unit, which should contain an entire picture. Do you happen to know what's the content of the access unit in which this crashes? Maybe that will shine light onto the problem.

My bet is on the SPS/PPS being sent on their own, separate from any slices.

In `decode_access_unit`, we can protect this with an if-let:

```
                let (picture, handle) = self.submit_picture()?;
                self.finish_picture(picture, handle)?;
```

> Maybe we can pass the picture as an argument to submit_picture, which would force the caller to take it itself and to check that it actually exists.

This can also work.

We cannot fail here, though:

```
            let new_field_picture = cur_field != prev_field;

            if new_field_picture {
                let (picture, handle) = self.submit_picture()?;
                self.finish_picture(picture, handle)?;
                self.handle_picture(timestamp, slice)?;
            }
```

That's because `handle_slice` is called after `handle_picture`, i.e.:

```
                    let slice = self.parser.parse_slice_header(nalu)?;
                    if self.cur_pic.is_none() {
                        self.handle_picture(timestamp, &slice)?;
                    }

                    self.handle_slice(timestamp, &slice)?;
```

**Comment by dwlsalmeida** [comment]
You can test the if-let change above (or similar), or wait for me to setup the chromebook, your call

**Comment by Gnurou** [comment]
I have a bunch of local changes needed to use ARCVM (notably to support DMABUF imports), I'll push them and try to make a release to crosvm can easily be built against them.

> My bet is on the SPS/PPS being sent on their own, separate from any slices.

That seems to be exactly what happens, after trying a workaround similar to what you suggested the crash did not happen anymore. I kept going and eventually implemented what I suggested above:

> Maybe we can pass the picture as an argument to submit_picture, which would force the caller to take it itself and to check that it actually exists.

I need to cleanup and rebase a bit, then I'll push that change. It removes quite a bunch of `unwrap`s, which is what we should want ultimately. 😄
